### ISSUE TITLE
a wml tag [query_location] that queries a location from the user

### ISSUE THREAD

**Issue by gfgtdf** [wesnoth/wesnoth#2325]
migrated from easycoding, no my own idea:

> * [query_location] - queries a location from the user
> ** variable
> ** [filter_location]
> ''Regarding [query_location]: probably a better name would be choose_location. This is used when a wml event is running and needs to get a target location hex from the user. For example, you right click on a catapult unit and choose the menu item "fire Catapult." Now a [message] says to click on any target within 5 hexes. Now [choose_location] is encountered and the mouse cursor changes to a target sign (possibly customizable). If the user's mouse drifts outside of the allowed locations it changes to a cancel icon. If the operation is cancelled then the variable is cleared and no location stored. --Sapient''
>
>''This can be implemented in lua by using wesnoth.game_events.on_mouse_action and wesnoth.syncronize_choice''

**Comment by CelticMinstrel** [comment]
Wait, we actually have the ability to do this now within the Lua API?

**Comment by gfgtdf** [comment]
we should, we have wesnoth.game_events.on_mouse_action whihc basically fires whenever a mode is slicked anywhere (similar to select events, but also works on on-unit hexed) and sync_choice to send it to others.

**Comment by ProditorMagnus** [comment]
wesnoth.game_events.on_mouse_action has not been documented anywhere. Trying to find out how it works.

**Comment by ProditorMagnus** [comment]
Seems it is fired on leftclick on hex with 1.13.x+.

**Comment by gfgtdf** [comment]
> wesnoth.game_events.on_mouse_action has not been documented anywhere. Trying to find out how it works.

it migth be buggy though, in paricular it'S current implementation might nreak if a user reassings the mouse1/mouse2 hotkeys in the hotkey menu.

**Comment by ProditorMagnus** [comment]
Adding extra select unit hotkey (to mouse1) makes it so that both ways fire this.

**Comment by gfgtdf** [comment]
ok i was bored so i made a sample implementation

```
function wesnoth.wml_actions.query_location(cfg)
	local res = wesnoth.synchronize_choice("location", function()
		local old_callback = wesnoth.game_events.on_mouse_action
		local xy = nil

		function wesnoth.game_events.on_mouse_action(x,y)

			if not xy then
				xy = { x = x, y = y }
			end

			if old_callback then
				--TODO: do we really want this?
				old_callback(x ,y)
			end
		end

		while not xy do
			while not xy do
				wesnoth.delay(10)
			end
			
			local result = helper.get_user_choice({ speaker = "narrator", message = "do you want to use location (" .. xy.x .. "," .. xy.y .. ")?" }, { "Yes", "No" })
			if result == 2 then
				xy = nil
			end
		end
		
		wesnoth.game_events.on_mouse_action = old_callback
	end)
end

```

**Comment by gfgtdf** [comment]
> Adding extra select unit hotkey (to mouse1) makes it so that both ways fire this.

What i mean is, that it currently onyl firex on the "select/move/attack" hotkey but on the "select" hotkey no the "move/attack" hotkey

**Comment by CelticMinstrel** [comment]
Will that allow selecting either empty or occupied hexes? Can we have some kind of hiliting of which hexes can be selected? (Might need new Lua API for that?)

**Comment by gfgtdf** [comment]
> Will that allow selecting either empty or occupied hexes?

yes, it wok on all hexes, as i said eariler a optional to add a location filter shoudl be added

> Can we have some kind of hiliting of which hexes can be selected? (Might need new Lua API for that?)

yes, i think wesnoth.add_tile_overlay would work here

**Comment by CelticMinstrel** [comment]
I was thinking of using the built-in method of hiliting where you can move by "graying" out the inaccessible hexes. Is that possible with Lua?

**Comment by gfgtdf** [comment]
well, i didn't testit but the obviuls attempt woudl be to use for example `wesnoth.add_tile_overlay(  .. terrain/darken.png .. )` on all hexes that do not match the filter.

**Comment by CelticMinstrel** [comment]
Is that also easily reversible? Isn't `add_tile_overlay` the API call used by `[item]`?

**Comment by ProditorMagnus** [comment]
Added to wiki what I learned about wesnoth.game_events.on_mouse_action here, someone might want to improve it.

**Comment by ProditorMagnus** [comment]
Going to work on this.

**Comment by ProditorMagnus** [comment]
Not actively working on it, but might continue if someone requests that. Current state posted to https://forums.wesnoth.org/viewtopic.php?f=12&t=47763.

**Comment by ProditorMagnus** [comment]
When player who is selecting location leaves game while there is another query_location in execution queue, whoever gets control of leaves side will become stuck. If there is no queued request, then it is possible to fail normally.

**Comment by ProditorMagnus** [comment]
https://forums.wesnoth.org/viewtopic.php?p=623198#p623198 should be good enough for singleplayer use. In mp I have been getting too many problems to find it viable to continue with this approach.

**Comment by gfgtdf** [comment]
ok after testing ravannas solution i think that `wesnoth.add_tile_overlay` is _not_ good enough to implement this, we need to eigher make `wesnoth.add_tile_overlay` more powerful, (beeing able to specify the layer on which in image is drawn on for example ) or add a new similar function.

**Comment by CelticMinstrel** [comment]
Maybe add support for multiple locations to `wesnoth.highlight_hex`?

I think we also need more power on the input detection side. For example, the ability to detect left and right clicks, or arbitrary keypresses.
